### ISSUE TITLE
confused about the function `forward_chop`

### ISSUE THREAD

**Issue by cmhungsteve** [huawei-noah/Pretrained-IPT#15]
Thank you for providing the wonderful codes.
However, I am confused about the function `forward_chop` in `model/__init.py`.
It seems that this function unfolds the input image into several patches and then feeds those patches into the IPT model, but I didn't find any detailed explanation in the paper or as comments in the code.
For example, what does `shave` mean here?
If I want to unfold the input image into non-overlapped patches, how should I do it?

Thank you.

**Comment by HantingChen** [comment]
In paper, we mentioned that "During the test, we crop the images in the test set into 48 × 48 patches with a 10 pixels overlap."  The reason is that our transformer model can only handle images with fixed input shape. 

`shave`  means the overlap pixel, you can set shave=0 to unfold image into non-overlapped patches for test, but the performance may drop a little.

**Comment by cmhungsteve** [comment]
from the code 153-161:
> padsize = int(self.patch_size)
    shave = int(self.patch_size/2)
    scale = self.scale[self.idx_scale]
    h_cut = (h-padsize)%(int(shave/2))
    w_cut = (w-padsize)%(int(shave/2))
    x_unfold = torch.nn.functional.unfold(x, padsize, stride=int(shave/2)).transpose(0,2).contiguous()

Let's say `padsize` is 48.
Doesn't that mean `shave` is 24 and `stride` is 12?
I am also confused by the relation between `shave` and `stride`.

If we set `shave` to 0, then `stride` will be 0, which is still weird......

**Comment by HantingChen** [comment]
Sorry for the misleading. I just found that we upload another version for chop (performance is slightly higher). 

So in this version, the height and width of the patch will be divide by 4 (12 for 48*48 inputs), and 12 pixels in the edge will be dropped, then 24*24 patches will be folded with a 12 overlap. 

If you want the original version (e.g., the code to unfold the input image into non-overlapped patches), we can upload it as an option.

**Comment by cmhungsteve** [comment]
Thank you for the prompt reply.
If I want to modify the current code into a "non-overlapped unfold/fold" version, how should I do that?
If you can upload a code for this version, that would be perfect!

Thank you.

**Comment by HantingChen** [comment]
You can find that in https://github.com/huawei-noah/Pretrained-IPT/blob/a16dd86e604302930d06d467384f143b494de557/model/__init__.py#L251

**Comment by cmhungsteve** [comment]
Thank you for sharing!
Another question is:
In lines 193-196 in `forward_chop`, there is a piece of codes like follows:
> y_ones = torch.ones(y_inter.shape, dtype=y_inter.dtype)
    divisor = torch.nn.functional.fold(torch.nn.functional.unfold(y_ones, padsize*scale-shave*scale, stride=int(shave/2\*scale)),((h-h_cut-shave)\*scale,(w-w_cut-shave)\*scale), padsize\*scale-shave\*scale, stride=int(shave/2\*scale))    
    y_inter = y_inter/divisor

I checked the PyTorch official website, it seems that the codes are used because unfold and fold are not invertible,

However, in `forward_chop_new`, there are no such kinds of codes.
Is there any reason?

Thank you.

**Comment by HantingChen** [comment]
In forward_chop, the unfold patches have overlaps. (More specifically, we not only cut pixels in the edges and merge the cutted images). While in forward_chop_new, we cut the pixels in the edges directly and each cutted patch do not have overlaps.

**Comment by cmhungsteve** [comment]
I see. Thank you for the explanation.

https://github.com/huawei-noah/Pretrained-IPT/blob/a16dd86e604302930d06d467384f143b494de557/model/__init__.py#L261
But why can't we set `stride`=`padsize`, but set `stride`=`padsize-shave` instead?
Doesn't that mean there are still overlaps between 48*48 blocks?

**Comment by HantingChen** [comment]
Yes, there are overlaps, but we will cut the pixels in the edges:

https://github.com/huawei-noah/Pretrained-IPT/blob/a16dd86e604302930d06d467384f143b494de557/model/__init__.py#L289

**Comment by cmhungsteve** [comment]
what happens if we unfold the image into non-overlapping 48*48 blocks?
If it's doable, how should I modify the codes?
Thank you.

**Comment by HantingChen** [comment]
> what happens if we unfold the image into non-overlapping 48*48 blocks?
> If it's doable, how should I modify the codes?
> Thank you.

Just set shave=0 in forward_chop_new

**Comment by cmhungsteve** [comment]
cool. Thank you.
What's the drawback if we set `shave`=0 (e.g., performance drops a lot)?

**Comment by HantingChen** [comment]
Yes, the performance will drops a lot (since the pixel in the edge will perform bad). Besides, the transition between different patches will be sharp and uneven.

**Comment by cmhungsteve** [comment]
Got it. That's clearer.
Thank you so much!

One last question:
It seems that the final output `y` is the combination of `y_inter`, `y_h_cut`, `y_w_cut`, `y_h_top`, `y_h_top`, `y_hw_cut`.
However, there are overlaps between them.
How do you deal with the overlaps?
Are there any details explained in the paper?

Thank you.

**Comment by HantingChen** [comment]
The edge of each patch is cutted, so when we put them together, the edge of the whole image is also cutted. That's the reason why we calculate y_h_cut, y_w_cut, y_h_top, y_h_top, y_hw_cut can put them into the edge of the whole image. Besides, the size of whole image may not be an integral multiple of patch size, so there must be some overlap (we put them in y_h_cut, y_w_cut).

If you set shave=0, it is not necesarry to calculate some of them. But you can also use this code since the final output is exactly the same.

**Comment by cmhungsteve** [comment]
I am trying `forward_chop_new` with `shave`=0.
And I thought in this line:
https://github.com/huawei-noah/Pretrained-IPT/blob/a16dd86e604302930d06d467384f143b494de557/model/__init__.py#L286
`y[...,:padsize*scale,:]` and `y_h_top` should be the same so that lines 286 & 287 are not necessary.
However, I found that `y[...,:padsize*scale,:]` and `y_h_top` are slightly different.
(i.e., cropping then feeding to IPT != feeding to IPT then cropping)

Do you know why this happens?

Thank you.

**Comment by HantingChen** [comment]
> I am trying `forward_chop_new` with `shave`=0. (i.e., there are exactly no overlaps)
> And I thought in this line:
> https://github.com/huawei-noah/Pretrained-IPT/blob/a16dd86e604302930d06d467384f143b494de557/model/__init__.py#L286
> 
> 
> `y[...,:padsize*scale,:]` and `y_h_top` should be the same so that lines 286 & 287 are not necessary.
> However, I found that `y[...,:padsize*scale,:]` and `y_h_top` are slightly different.
> (i.e., cropping then feeding to IPT != feeding to IPT then cropping)
> Do you know why this happens?
> 
> Thank you.

That's strange. Maybe the nn.layernorm is calculated differently when feeding with different batches. I think this will not affect its performance.

**Comment by cmhungsteve** [comment]
I also have a question about the inputs that you feed into the Transformer:
https://github.com/huawei-noah/Pretrained-IPT/blob/a16dd86e604302930d06d467384f143b494de557/model/__init__.py#L281
Assume `x_unfold` has the shape `[25, 3, 48, 48]` (25 patches, each one has resolution 48x48).
And assume after the head encoder, the input becomes `[25, 32, 48, 48]`.

The part that makes me confused is that there is another unfold function before the Transformer encoder:
https://github.com/huawei-noah/Pretrained-IPT/blob/a16dd86e604302930d06d467384f143b494de557/model/ipt.py#L137
Since `self.patch_dim`=3, this makes `x` becomes `[(16*16), 25, (32*3*3)]` and become the inputs of the multi-head attention.
This means that the sequence length of attention is 256 here, not 25.
Does that mean the attention mechanism is not used to learn the relation between 25 patches, but learn the relation between pixels inside each patch instead?
If so, this seems different from the original ViT paper, which uses attention to learn the relation between patches inside an image.
Why is this difference here?

Thank you.

**Comment by HantingChen** [comment]
I think there might be some misunderstanding. 

The inputs of our IPT model is exactly 48*48*3 images (whose scale is 224*224*3 in ViT). And the 48*48 images will be cropped into 3*3 small patches with 16*16 seq_length. 

The fowrard_chop function is used for handling different input size, since our IPT can only take 48*48 size images.

**Comment by cmhungsteve** [comment]
I see. Now it's much clearer.
I also want to check if I understand the paper correctly.
In Section 3.1, H and W are both 48, and P is 3. Is that correct?

**Comment by cmhungsteve** [comment]
Got it. Thank you so much!!!
